File: src/events.ts
/**
 * ARS principal-track events.
 *
 * Telaro's ARS-Solana implementation is **event-sourced**, mirroring the
 * ARS abstract layer: a job carries no mutable state of its own — its
 * lifecycle state is derived by replaying this append-only event log
 * (see `state.ts`).
 *
 * Why event-sourced (this resolves the open question in
 * docs/internal/ars-dual-track-redesign.md §7): it matches the ARS
 * reference design, makes the principal-track lifecycle auditable, and
 * lets on-chain telaro events (bond, slash, credit draw) and off-chain
 * ARS decisions (the underwriting verdict) feed one ordered log.
 *
 * The principal track is the only track Telaro implements — Telaro is a
 * capital-risk layer, not a generic service-fee escrow (dual-track
 * redesign §4.2).
 */

/** Fields every event carries. */
interface BaseEvent {
  /** ARS job id. On Solana this is the telaro `action_hash` (§4.1). */
  jobId: string;
  /** Unix seconds the event was recorded. */
  at: number;
}

/** A capital-handling job is opened by a DApp delegating to an agent. */
export interface JobOpened extends BaseEvent {
  type: "JobOpened";
  /** Agent wallet — the Telaro agent key / x402 transfer authority. */
  agent: string;
  /** DApp / capital provider that opened the job (ARS Requestor). */
  requestor: string;
  /** Capital the agent will handle for this job (USDC atomic, 6dp). */
  exposureAtomic: bigint;
}

/** Underwriting review has begun (the telaro `view_bond` evaluation). */
export interface UnderwritingStarted extends BaseEvent {
  type: "UnderwritingStarted";
}

/** The underwriting verdict — `view_bond` passed or failed. */
export interface UnderwritingDecided extends BaseEvent {
  type: "UnderwritingDecided";
  passed: boolean;
  /** Gate failure code when `passed` is false (NOT_BONDED, FROZEN, …). */
  failureCode?: string;
}

/** Principal was released to the agent (a `request_credit` draw). */
export interface PrincipalReleased extends BaseEvent {
  type: "PrincipalReleased";
  amountAtomic: bigint;
}

/** The agent's execution result was recorded (telaro `record_action`). */
export interface EvidenceSubmitted extends BaseEvent {
  type: "EvidenceSubmitted";
  /** telaro `action_hash` of the recorded action. */
  actionHash: string;
  outcome: "success" | "failed";
}

/** A claim was filed against the job — hands off to the telaro claim flow. */
export interface Disputed extends BaseEvent {
  type: "Disputed";
  /** telaro `Claim` PDA. */
  claim: string;
}

/** The job reached a terminal outcome. */
export interface Closed extends BaseEvent {
  type: "Closed";
  resolution:
    | "no_dispute" // dispute window expired clean
    | "dispute_upheld" // claim resolved against the agent (bond slashed)
    | "dispute_rejected"; // claim resolved in the agent's favour
}

export type ArsEvent =
  | JobOpened
  | UnderwritingStarted
  | UnderwritingDecided
  | PrincipalReleased
  | EvidenceSubmitted
  | Disputed
  | Closed;

export type ArsEventType = ArsEvent["type"];

File: python/src/telaro_ars/events.py
"""
ARS principal-track events for the Telaro Solana implementation.

These mirror `src/events.ts` from `@telaro/ars-solana` byte for byte
(same fields, same names, same semantics). A Python consumer that
holds an event log produced by either side can replay it through
`state.replay`.

Job lifecycle is event-sourced: a `Job` carries no mutable state of
its own. The current state is derived by folding this log through
`apply_event` (state.py). Event-sourced design matches the ARS
reference and makes the principal-track lifecycle auditable.

Scope: principal track only. Fee track is deferred to v0.2 per
SPEC.md sections 1 and 10.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Union


@dataclass(frozen=True)
class JobOpened:
    """A capital-handling job is opened by a DApp delegating to an agent."""

    job_id: str
    at: int
    agent: str
    requestor: str
    exposure_atomic: int
    type: Literal["JobOpened"] = "JobOpened"


@dataclass(frozen=True)
class UnderwritingStarted:
    """Underwriting review has begun (the Telaro `view_bond` evaluation)."""

    job_id: str
    at: int
    type: Literal["UnderwritingStarted"] = "UnderwritingStarted"


@dataclass(frozen=True)
class UnderwritingDecided:
    """The underwriting verdict from `view_bond`."""

    job_id: str
    at: int
    passed: bool
    failure_code: str | None = None
    type: Literal["UnderwritingDecided"] = "UnderwritingDecided"


@dataclass(frozen=True)
class PrincipalReleased:
    """Principal was released to the agent (a `request_credit` draw)."""

    job_id: str
    at: int
    amount_atomic: int
    type: Literal["PrincipalReleased"] = "PrincipalReleased"


@dataclass(frozen=True)
class EvidenceSubmitted:
    """The agent's execution result was recorded (Telaro `record_action`)."""

    job_id: str
    at: int
    action_hash: str
    outcome: Literal["success", "failed"]
    type: Literal["EvidenceSubmitted"] = "EvidenceSubmitted"


@dataclass(frozen=True)
class Disputed:
    """A claim was filed against the job. Hands off to the Telaro claim flow."""

    job_id: str
    at: int
    claim: str
    type: Literal["Disputed"] = "Disputed"


@dataclass(frozen=True)
class Closed:
    """The job reached a terminal outcome."""

    job_id: str
    at: int
    resolution: Literal["no_dispute", "dispute_upheld", "dispute_rejected"]
    type: Literal["Closed"] = "Closed"


ArsEvent = Union[
    JobOpened,
    UnderwritingStarted,
    UnderwritingDecided,
    PrincipalReleased,
    EvidenceSubmitted,
    Disputed,
    Closed,
]
